graph-builder/src/registry.rs

```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
+            }
+        })
+        .and_then(|dclient| {
+            dclient.is_auth(None).and_then(move |is_auth| {
+                if is_auth {
+                    Err("no login performed, but already authenticated".into())
+                } else {
+                    Ok(dclient)
+                }
+            })
+        })
+        .and_then(move |dclient| {
+            dclient.login(&[&login_scope]).and_then(move |token| {
+                dclient
+                    .is_auth(Some(token.token()))
+                    .and_then(move |is_auth| {
+                        if !is_auth {
+                            Err("login failed".into())
+                        } else {
+                            Ok(dclient.set_token(Some(token.token())))
+                        }
+                    })
+            })
+        })
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
-    let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(|| -> Result<Box<dyn std::io::Read>, Error> {
+            match credentials_path {
+                Some(path) => match File::open(path) {
+                    Ok(file) => Ok(Box::new(file)),
+                    Err(e) => Err(e.into()),
+                },
+                None => Ok(Box::new(std::io::empty())),
+            }
+        }()?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let authenticated_client = tcore
+        .run(authenticate_client(
+            &mut client,
+            format!("repository:{}:pull", &repo),
+        ))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let tags = match tcore.run(
+        authenticated_client
+            // According to https://docs.docker.com/registry/spec/api/#listing-image-tags
+            // the tags should be ordered lexically but they aren't
+            .get_tags(&repo, Some(20))
+            .map_err(|e| format_err!("{}", e))
+            .collect(),
+    ) {
+        Ok(mut tags) => {
+            if tags.len() == 0 {
+                bail!("{}/{} has no tags", registry, repo);
+            }
+            tags.sort();
```

Comment:
I designed the tests in a way to expect a certain vector of results because the documentation says the tags are returned in a lexical sorted order. Quay seems to diverge from that so this sort enforces what the docker registry API documents.

Reply:
We should really try to avoid coupling to Quay as much as possible. CI cannot depend on it being up and they have already expressed concern over the amount of load we are putting on them. I'd rather we just adjusted our tests instead of the product.

Reply:
It's possible to neglect the order in the tests, with the downside of making a failed test more difficult to debug.